### PR TITLE
TEST: introduce unit test for REST placeholders

### PR DESCRIPTION
**Pull request by jimirocks**

### CODE REVIEW

File: lib/gooddata/rest/phmap.rb [hunk 1]
```diff
@@ -7,8 +7,16 @@
 module GoodData
   module Rest
     class Connection
+
+      def self.map_placeholders(title)
+        @@PH_MAP.each do |pm|
```
Comment (yenkins) [inline lib/gooddata/rest/phmap.rb]: Declares the class variable '@@ph_map' (ClassVariable)

File: spec/unit/rest/phmap_spec.rb [hunk 2]
```diff
@@ -0,0 +1,104 @@
+# encoding: UTF-8
+#
+# Copyright (c) 2010-2017 GoodData Corporation. All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+require 'gooddata/rest/phmap'
+
+class PhEx < Struct.new(:title, :mapped)
```
Comment (yenkins) [inline spec/unit/rest/phmap_spec.rb]: Has no descriptive comment (IrresponsibleModule)

Comment (yenkins) [inline spec/unit/rest/phmap_spec.rb]: Style/StructInheritance: Don't extend an instance initialized by `Struct.new`.

File: lib/gooddata/rest/phmap.rb [hunk 3]
```diff
@@ -7,8 +7,16 @@
 module GoodData
   module Rest
     class Connection
+
```
Comment (yenkins) [inline lib/gooddata/rest/phmap.rb]: Style/EmptyLinesAroundClassBody: Extra empty line detected at class body beginning.

File: lib/gooddata/rest/phmap.rb [hunk 4]
```diff
@@ -7,8 +7,16 @@
...
+      def self.map_placeholders(title)
+        @@PH_MAP.each do |pm|
+          break if title.gsub!(pm[1], pm[0])
+        end
+        title
+      end
+
       # PH_MAP for wildcarding of URLs in reports
-      PH_MAP = [
+      @@PH_MAP = [
```
Comment (yenkins) [inline lib/gooddata/rest/phmap.rb]: Style/ClassVars: Replace class var @@PH_MAP with a class instance var.

File: spec/unit/rest/phmap_spec.rb [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+      PhEx.new('/gdc/dataload/projects/uif55z7zg9zunnsffsqu6nzyg38j20qg/outputStage', '/gdc/dataload/projects/{id}/outputStage'),
+
+      PhEx.new('/gdc/datawarehouse/instances/c90ef89f07978760523dd409165c71f3', '/gdc/datawarehouse/instances/{id}'),
+      PhEx.new('/gdc/datawarehouse/executions/exec123', '/gdc/datawarehouse/executions/{id}'),
+
+      PhEx.new('/gdc/domains/stg2-lcm-prod/dataproducts/MY-DP/clients?segment=MY-SEG4', '/gdc/domains/{id}/dataproducts/{data_product}/clients?segment={segment}'),
+      PhEx.new('/gdc/domains/MY_DOMAIN3/dataproducts/4_7-prod/clients/some-client1', '/gdc/domains/{id}/dataproducts/{data_product}/clients/{client}'),
+      PhEx.new('/gdc/domains/D_-3/dataproducts/DP_1-/clients', '/gdc/domains/{id}/dataproducts/{data_product}/clients'),
+
+      PhEx.new('/gdc/domains/d2/dataproducts/dp-6/segments/seg_1/synchronizeClients/results/_-res8/details?offset=0&limit=1000', '/gdc/domains/{id}/dataproducts/{data_product}/segments/{segment}/synchronizeClients/results/{result}/details?offset={offset}&limit={limit}'),
```
Comment (yenkins) [inline spec/unit/rest/phmap_spec.rb]: Metrics/LineLength: Line is too long. [271/180]

File: spec/unit/rest/phmap_spec.rb [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+      PhEx.new('/gdc/domains/D/dataproducts/DP/segments/SEG/synchronizeClients/results/RES', '/gdc/domains/{id}/dataproducts/{data_product}/segments/{segment}/synchronizeClients/results/{result}'),
```
Comment (yenkins) [inline spec/unit/rest/phmap_spec.rb]: Metrics/LineLength: Line is too long. [197/180]

File: spec/unit/rest/phmap_spec.rb [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+      PhEx.new('/gdc/domains/D/dataproducts/DP/segments/SEG', '/gdc/domains/{id}/dataproducts/{data_product}/segments/{segment}'),
+      PhEx.new('/gdc/domains/D/dataproducts/DP/segments', '/gdc/domains/{id}/dataproducts/{data_product}/segments'),
+
+      PhEx.new('/gdc/domains/D/dataproducts/DP/provisionClientProjects/results/RES/details?offset=7&limit=50', '/gdc/domains/{id}/dataproducts/{data_product}/provisionClientProjects/results/{result}/details?offset={offset}&limit={limit}'),
```
Comment (yenkins) [inline spec/unit/rest/phmap_spec.rb]: Metrics/LineLength: Line is too long. [239/180]

File: spec/unit/rest/phmap_spec.rb [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+      PhEx.new('/gdc/domains/D/dataproducts/DP/provisionClientProjects/results/RES', '/gdc/domains/{id}/dataproducts/{data_product}/provisionClientProjects/results/{result}'),
+      PhEx.new('/gdc/domains/D/dataproducts/DP/provisionClientProjects', '/gdc/domains/{id}/dataproducts/{data_product}/provisionClientProjects'),
+      PhEx.new('/gdc/domains/D/dataproducts/DP/updateClients?deleteExtraInSegments=SEG', '/gdc/domains/{id}/dataproducts/{data_product}/updateClients?deleteExtraInSegments={segments}'),
```
Comment (yenkins) [inline spec/unit/rest/phmap_spec.rb]: Metrics/LineLength: Line is too long. [185/180]

File: spec/unit/rest/phmap_spec.rb [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+      PhEx.new('/gdc/domains/D/dataproducts/DP/updateClients', '/gdc/domains/{id}/dataproducts/{data_product}/updateClients'),
+
+      PhEx.new('/gdc/domains/D/dataproducts/DP', '/gdc/domains/{id}/dataproducts/{data_product}'),
+      PhEx.new('/gdc/domains/D/dataproducts', '/gdc/domains/{id}/dataproducts'),
+
+      PhEx.new('/gdc/domains/D/segments/SEG/synchronizeClients/results/RES/details?offset=0&limit=100', '/gdc/domains/{id}/segments/{segment}/synchronizeClients/results/{result}/details?offset={offset}&limit={limit}'),
```
Comment (yenkins) [inline spec/unit/rest/phmap_spec.rb]: Metrics/LineLength: Line is too long. [218/180]

File: spec/unit/rest/phmap_spec.rb [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+      PhEx.new('/gdc/domains/D/segments/SEG/synchronizeClients/results/RES', '/gdc/domains/{id}/segments/{segment}/synchronizeClients/results/{result}'),
+      PhEx.new('/gdc/domains/D/segments/SEG/', '/gdc/domains/{id}/segments/{segment}/'),
+      PhEx.new('/gdc/domains/D/segments/SEG', '/gdc/domains/{id}/segments/{segment}'),
+
+      PhEx.new('/gdc/domains/D/clients?segment=SEG', '/gdc/domains/{id}/clients?segment={segment}'),
+      PhEx.new('/gdc/domains/D/clients/C/settings/lcm.title', '/gdc/domains/{id}/clients/{client_id}/settings/lcm.title'),
+      PhEx.new('/gdc/domains/D/clients/C/settings/lcm.token', '/gdc/domains/{id}/clients/{client_id}/settings/lcm.token'),
+      PhEx.new('/gdc/domains/D/clients/C', '/gdc/domains/{id}/clients/{client_id}'),
+      PhEx.new('/gdc/domains/D/provisionClientProjects/results/R/details?offset=0&limit=100', '/gdc/domains/{id}/provisionClientProjects/results/{result}/details?offset={offset}&limit={limit}'),
```
Comment (yenkins) [inline spec/unit/rest/phmap_spec.rb]: Metrics/LineLength: Line is too long. [194/180]

File: spec/unit/rest/phmap_spec.rb [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+      PhEx.new('/gdc/domains/D/provisionClientProjects/results/R', '/gdc/domains/{id}/provisionClientProjects/results/{result}'),
+      PhEx.new('/gdc/domains/D/provisionClientProjects', '/gdc/domains/{id}/provisionClientProjects'),
+      PhEx.new('/gdc/domains/D/updateClients', '/gdc/domains/{id}/updateClients'),
+      PhEx.new('/gdc/domains/D/', '/gdc/domains/{id}/'),
+
+      PhEx.new('/gdc/exporter/result/a23f/c56', '/gdc/exporter/result/{id}/{id}'),
+
+      PhEx.new('/gdc/internal/lcm/domains/D/dataproducts/DP/segments/SEG/syncProcesses/P', '/gdc/internal/lcm/domains/{id}/dataproducts/{data_product}/segments/{segment}/syncProcesses/{process}'),
```
Comment (yenkins) [inline spec/unit/rest/phmap_spec.rb]: Metrics/LineLength: Line is too long. [196/180]

File: spec/unit/rest/phmap_spec.rb [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+      PhEx.new('/gdc/internal/lcm/domains/D/dataproducts/DP/segments/SEG/syncProcesses', '/gdc/internal/lcm/domains/{id}/dataproducts/{data_product}/segments/{segment}/syncProcesses'),
```
Comment (yenkins) [inline spec/unit/rest/phmap_spec.rb]: Metrics/LineLength: Line is too long. [184/180]

File: spec/unit/rest/phmap_spec.rb [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+      PhEx.new('/gdc/projects/PID/dataload/processes/PRID', '/gdc/projects/{id}/dataload/processes/{id}'),
+      PhEx.new('/gdc/projects/PID/', '/gdc/projects/{id}/'),
+      PhEx.new('/gdc/projects/PID', '/gdc/projects/{id}'),
+
+      PhEx.new('/gdc/userGroups/GID', '/gdc/userGroups/{id}'),
+
+      PhEx.new('/gdc/account/profile/UID', '/gdc/account/profile/{id}'),
+      PhEx.new('/gdc/account/login/UID', '/gdc/account/login/{id}'),
+      PhEx.new('/gdc/account/domains/D/users?login=rubydev+admin@gooddata.com', '/gdc/account/domains/{id}/users?login={login}'),
+      PhEx.new('/gdc/account/domains/D', '/gdc/account/domains/{id}'),
```
Comment (yenkins) [inline spec/unit/rest/phmap_spec.rb]: Style/TrailingCommaInLiteral: Avoid comma after the last item of an array.

File: spec/unit/rest/phmap_spec.rb [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    ].each do |example|
+      expect(GoodData::Rest::Connection::map_placeholders(example[:title])).to eq(example[:mapped])
```
Comment (yenkins) [inline spec/unit/rest/phmap_spec.rb]: Style/ColonMethodCall: Do not use `::` for method calls.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/814/ : SUCCESS in 14s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/681/ : FAILURE in 1m 23s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/942/ : FAILURE in 1m 10s

File: lib/gooddata/rest/phmap.rb (hunk 4 above)
Comment (panjan) [reply lib/gooddata/rest/phmap.rb]: A Ruby constant would be more suitable for storing the mapping.

File: spec/unit/rest/phmap_spec.rb [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+    PhEx.new('dataload process', '/gdc/projects/PID/dataload/processes/PRID', '/gdc/projects/{id}/dataload/processes/{id}'),
+    PhEx.new('project', '/gdc/projects/PID/', '/gdc/projects/{id}/'),
+    PhEx.new('project', '/gdc/projects/PID', '/gdc/projects/{id}'),
+
+    PhEx.new('userGroup', '/gdc/userGroups/GID', '/gdc/userGroups/{id}'),
+
+    PhEx.new('profile', '/gdc/account/profile/UID', '/gdc/account/profile/{id}'),
+    PhEx.new('login', '/gdc/account/login/UID', '/gdc/account/login/{id}'),
+    PhEx.new('domain user', '/gdc/account/domains/D/users?login=rubydev+admin@gooddata.com', '/gdc/account/domains/{id}/users?login={login}'),
+    PhEx.new('account domain', '/gdc/account/domains/D', '/gdc/account/domains/{id}'),
```
Comment (yenkins) [inline spec/unit/rest/phmap_spec.rb]: Style/TrailingCommaInLiteral: Avoid comma after the last item of an array.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/819/ : SUCCESS in 15s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/686/ : FAILURE in 39s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/947/ : FAILURE in 47s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/820/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/687/ : SUCCESS in 35s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/948/ : SUCCESS in 44s

File: spec/unit/rest/phmap_spec.rb [hunk 16]
```diff
@@ -0,0 +1,115 @@
+# encoding: UTF-8
+#
+# Copyright (c) 2010-2017 GoodData Corporation. All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+require 'gooddata/rest/phmap'
+
+PhEx = Struct.new(:name, :title, :mapped)
```
Comment (panjan) [inline spec/unit/rest/phmap_spec.rb]: Could you please rename `PhEx` to something more descriptive? E.g. `PlaceholderExample`

File: spec/unit/rest/phmap_spec.rb [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+    PlaceholderExample.new('domain segment', '/gdc/domains/D/segments/SEG', '/gdc/domains/{id}/segments/{segment}'),
+
+    PlaceholderExample.new('domain client segment', '/gdc/domains/D/clients?segment=SEG', '/gdc/domains/{id}/clients?segment={segment}'),
+    PlaceholderExample.new('client settings title', '/gdc/domains/D/clients/C/settings/lcm.title', '/gdc/domains/{id}/clients/{client_id}/settings/lcm.title'),
+    PlaceholderExample.new('client settings token', '/gdc/domains/D/clients/C/settings/lcm.token', '/gdc/domains/{id}/clients/{client_id}/settings/lcm.token'),
+    PlaceholderExample.new('domain client', '/gdc/domains/D/clients/C', '/gdc/domains/{id}/clients/{client_id}'),
+    PlaceholderExample.new('domain provisionClientProjects result details', '/gdc/domains/D/provisionClientProjects/results/R/details?offset=0&limit=100',
+                           '/gdc/domains/{id}/provisionClientProjects/results/{result}/details?offset={offset}&limit={limit}'),
+    PlaceholderExample.new('domain provisionClientProjects result', '/gdc/domains/D/provisionClientProjects/results/R', '/gdc/domains/{id}/provisionClientProjects/results/{result}'),
```
Comment (yenkins) [inline spec/unit/rest/phmap_spec.rb]: Metrics/LineLength: Line is too long. [182/180]

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/821/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/688/ : FAILURE in 40s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/949/ : SUCCESS in 36s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/822/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/689/ : SUCCESS in 34s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/950/ : SUCCESS in 34s

**Comment by panjan** [comment]
GJ! Thank you ❤️

**Comment by panjan** [comment]
ok to merge

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-merge-compliance-check/315/ : SUCCESS in 15s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/690/ : SUCCESS in 33s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/951/ : SUCCESS in 34s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1267/ : FAILURE in 1h 08m 11s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/1027/ : SUCCESS in 59m 47s

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-merge-compliance-check/321/ : SUCCESS in 21s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/757/ : SUCCESS in 43s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1025/ : SUCCESS in 43s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1297/ : FAILURE in 54m 49s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/1061/ : SUCCESS in 30m 26s

**Comment by jimirocks** [comment]
extended test

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/908/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1298/ : SUCCESS in 58m 59s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/1062/ : SUCCESS in 39m 07s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-merge-compliance-check/322/ : SUCCESS in 17s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/759/ : SUCCESS in 31s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1027/ : SUCCESS in 3m 21s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1300/ : SUCCESS in 56m 39s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/1064/ : SUCCESS in 32m 34s

**Pull request merged**